당첨이벤트 발생시 재고 감소 및 outbox테이블에 이벤트 저장기능

qnmgjeklnbrl opened the pull request:
- 낮은확률로 발생할 수 있는 재고 감소 동시요청이 들어올 때  낙관적락과 @retry로 제어합니다.
- 아웃박스 테이블에 결과 저장
- 추후 아웃박스 테이블에 결과 저장후 카프카에 결과 메시지 발행기능 과 아웃박스 스케줄러 기능 개발 예정입니다.

qnmgjeklnbrl committed:
chore: 의존성추가 및 패키지 구조 변경

qnmgjeklnbrl committed:
chore: kafka 설정

qnmgjeklnbrl committed:
feat: kafka에서 메시지 수신

qnmgjeklnbrl committed:
feat: outbox엔티티, auditing을 위한 base엔티티 추가

qnmgjeklnbrl committed:
feat: EventListener 인터페이스, 구현체 추가 및 재고 감소 기능

- 추후 확장성 고려해 인터페이스를 구현하는 형태로 개발
- 토픽에 발행된 이벤트를 수신하여 해당 상품 재고 감소

qnmgjeklnbrl committed:
chore: 패키지 구조 변경

qnmgjeklnbrl committed:
chore: retry 의존성 추가 및 설정클래스

qnmgjeklnbrl committed:
feat: 예외처리 추가

- 재고 감소시도시 수량이 없는경우 OutOfStockException 발생
- 동시경합상황시 재시도 5번 실패하면 ConcurrencyException 발생

qnmgjeklnbrl committed:
feat: 재고감소시 동시성 처리 및 outbox테이블에 결과저장

- 동시에 재고감소 요청이 들어오면 재시도
- 5번 까지 재시도후 실패시 예외 발생
- outbox테이블에 결과 저장

qnmgjeklnbrl committed:
feat: 재고감소시 동시성 처리 및 outbox테이블에 결과저장

- 동시에 재고감소 요청이 들어오면 재시도
- 5번 까지 재시도후 실패시 예외 발생
- outbox테이블에 결과 저장

qnmgjeklnbrl committed:
test: 동시성 제어 테스트

qnmgjeklnbrl committed:
refactor: 로깅 수정

dlehddn commented on prize/src/main/java/ddalkak/prize/dto/DecreaseResultEvent.java:
```diff
@@ -0,0 +1,8 @@
+public record DecreaseResultEvent(
+        Long eventId,
+        Long prizeId,
+        boolean result
```
result를 상수로 뒀으면 이해하기 편할거 같아요.

dlehddn commented on prize/src/main/java/ddalkak/prize/eventlistener/impl/KafkaEventListener.java:
```diff
@@ -0,0 +1,51 @@
+     */
+    @Override
+    @KafkaListener(
+            topics = "prize_decrease",
```
토픽명 규칙을 prize.decrease <- 이렇게 dot을 활용하도록 맞추면 좋을것 같습니다.

dlehddn commented on prize/src/main/java/ddalkak/prize/config/error/GlobalExceptionHandler.java:
```diff
@@ -52,6 +49,18 @@ protected ResponseEntity<ErrorResponse> handle(ConstraintViolationException e) {
         log.error("request 검증 실패", e);
        return createErrorResponseEntity(e.getMessage(), HttpStatus.BAD_REQUEST);
     }
+    @ExceptionHandler(ConcurrencyException.class)
```
Kafka 전용 에러 핸들러가 별도로 있네요. 찾아보세요 한번ㅎ

dlehddn reviewed (commented):
이야기 나눈 retry 관련 제외하고 코멘트 드렸습니다.
고생하셨습니다^^ 수정하고 연락주세요.

qnmgjeklnbrl committed:
feat: 요청사항 반영 수정

- KafkaConsumerConfig: 수동커밋을 위한 설정 변경, DefaultErrorHandler 에서 재시도 실패 시 outbox테이블에 결과 저장하도록 구현
- 패키지구조 변경: 기존 EventListener 인터페이스를 이벤트 수신만 하는 KafkaEventListener, 비즈니스로직 담당하는 EventHandler로 분리
- KafkaEventListener @Transactional과 수동커밋 사용으로 변경
- outbox테이블에 재고감소 결과를 boolean이 아닌 enum타입으로 변경
- topic명 규약에 맞게 변경
- 필요없는 예외 및 핸들러 삭제

qnmgjeklnbrl committed:
test: 동일한 상품에 재고 감소 요청 상황 테스트

qnmgjeklnbrl committed:
fix: 트랜잭션 구조 변경, 상품 재고 없을때 처리 방식 변경

qnmgjeklnbrl committed:
fix: 트랜잭션 구조 변경, 상품 재고 없을때 처리 방식 변경

qnmgjeklnbrl committed:
refactor: 자잘한 코드 정리

dlehddn reviewed (approved):
고생하셨습니다ㅎ
